src/org/rascalmpl/library/analysis/diff/edits/HiFiLayoutDiff.rsc

```diff
@@ -0,0 +1,118 @@
+* if `originalTree !:= formattedTree` the algorithm will produce junk. It will break the syntactical correctness of the source code and forget source code comments.
+}
+@benefits{
+* Recovers source code comments which have been lost during earlier steps in the formatting pipeline. This makes losing source code comments an independent concern of a declarative formatter.
+* Recovers the original capitalization of case-insensitive literals which may have been lost during earlier steps in the formatting pipeline.
+* Is agnostic towards the design of earlier steps in the formatting pipeline, so lang as `formattedTree := originalTree`. This means that
+the pipeline may change layout (whitespace and comments and capitalization of case-insensitive literals), but nothing else.
+}
+module analysis::diff::edits::HiFiLayoutDiff
+
+extend analysis::diff::edits::HiFiTreeDiff;
+import ParseTree; // this should not be necessary because imported by HiFiTreeDiff
+import String; // this should not be be necessary because imported by HiFiTreeDiff
+
+@synopsis{Extract TextEdits for the differences in whitespace between two otherwise identical ((ParseTree))s.}
+@description{
+See ((HiFiLayoutDiff)).
+}
+list[TextEdit] layoutDiff(Tree original, Tree formatted, bool copyComments = true) {
+    assert original := formatted : "nothing except layout and keyword fields may be different for layoutDiff to work correctly.";
+
+    @synopsis{rec is the recursive workhorse, doing a pairwise recursion over the original and the formatted tree}
+    @description{
+        We recursively skip over every "equal" pairs of nodes, until we detect two different _layout_ nodes. The original location
+        of that node and the new contents of the formatted node is used to construct a replace ((TextEdit)), and 
+        optionally the original layout is inspected for source code comments which may have been lost. Literals are skipped 
+        explicitly to avoid arbitrary edits for case insensitive literals, and to safe some time.
+    }
+
+    // if layout differences are detected, here we produce a `replace` node:
+    list[TextEdit] rec(
+        t:appl(prod(Symbol tS, _, _), list[Tree] tArgs), // layout is not necessarily parsed with the same rules (i.e. comments are lost!)
+        u:appl(prod(Symbol uS, _, _), list[Tree] uArgs))
+        = [replace(t@\loc, copyComments ? learnComments(t, "<u>") : "<u>") | tArgs != uArgs] 
+        when 
+            delabel(tS) is layouts, 
+            delabel(uS) is layouts,
+            tArgs != uArgs; 
+
+    // matched literal trees generate empty diffs
+    list[TextEdit] rec(
+        appl(prod(lit(_), _, _), list[Tree] _),
+        appl(prod(lit(_), _, _), list[Tree] _))
+        = [];
+
+    // matched case-insensitive literal trees generate empty diffs such that the original is maintained
+    list[TextEdit] rec(
+        appl(prod(cilit(_), _, _), list[Tree] _),
+        appl(prod(cilit(_), _, _), list[Tree] _))
+        = [];
+
```

Comment:
I rewrote to new base cases (indeed missing!) for efficiency. Because the trees are structurally equal,
we may assume every char and every cycle that is matched will be structurally equal and we don't have to test equality.